### PR TITLE
refactor file I/O implemntation

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <kpango@vdaas.org>

### Description:
This PR includes 3 changes
1. remove std io.Copy and replace it to internal/io.Copy which uses auto scale pool buffer
```
❯ go test -bench . -run=NONE -count=1 --benchmem
goos: linux
goarch: amd64
pkg: github.com/vdaas/vald/internal/io
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkStandardIOCopy-16                        411262              3579 ns/op           32816 B/op          3 allocs/op
BenchmarkStandardIOCopyBuffer-16                  366306              3564 ns/op           32816 B/op          3 allocs/op
BenchmarkValdIOCopy-16                           3392606               360.3 ns/op            48 B/op          2 allocs/op
BenchmarkStandardIOCopyParallel-16                439708              2668 ns/op           32816 B/op          3 allocs/op
BenchmarkStandardIOCopyBufferParallel-16          494727              2345 ns/op           32816 B/op          3 allocs/op
BenchmarkValdIOCopyParallel-16                  19977921                65.29 ns/op           50 B/op          2 allocs/op
PASS
ok      github.com/vdaas/vald/internal/io       9.137s
```
2. use *os.File.Sync() for stability
3. grpc Pool error messages refactor

### Environment:

- Go Version: 1.16.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/config/config.go [hunk 1]
```diff
@@ -72,13 +73,15 @@ func Read(path string, cfg interface{}) (err error) {
...
+	switch ext := filepath.Ext(path); ext {
+	case ".yaml", ".yml":
 		err = yaml.NewDecoder(f).Decode(cfg)
 	case ".json":
 		err = json.Decode(f, cfg)
+	default:
+		return errors.ErrUnsupportedConfigFileType(ext)
 	}
-	return err
+	return nil
```
Comment (kevindiu) [inline internal/config/config.go:84]: why do we return nil here? 🤔

Comment (kpango) [reply internal/config/config.go:84]: I think this statement is unreachable.

Comment (kevindiu) [reply internal/config/config.go:84]: then do we need to change line 78 & line 80 to return?

Comment (kpango) [reply internal/config/config.go:84]: I think either will work.

File: internal/io/copy.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+func New() Copier {
+	c := new(copier)
+	atomic.StoreInt64(&c.bufSize, int64(defaultBufferSize))
+	c.pool = sync.Pool{
+		New: func() interface{} {
+			return bytes.NewBuffer(make([]byte, int(atomic.LoadInt64(&c.bufSize))))
+		},
+	}
+	for i := 0; i < 10; i++ {
```
Comment (kevindiu) [inline internal/io/copy.go:51]: can you tell me why do we initialize 10 buffers for a single copier?

**Review by hlts2 — DISMISSED** [review]

**Review by kevindiu — DISMISSED** [review]

File: internal/db/storage/blob/s3/reader/reader.go [hunk 3]
```diff
@@ -66,7 +66,7 @@ func New(opts ...Option) (Reader, error) {
 	}
 
 	if r.ctxio == nil {
-		return nil, errors.NewErrInvalidOption("ctxio", r.ctxio)
+		return nil, errors.NewErrInvalidOption("txio", r.ctxio)
```
Comment (kevindiu) [inline internal/db/storage/blob/s3/reader/reader.go:69]:
```suggestion
		return nil, errors.NewErrInvalidOption("ctxio", r.ctxio)
```

**Review by kevindiu — DISMISSED** [review]
LGTM

**Review by hlts2 — APPROVED** [review]

**Pull request closed without merging**